### PR TITLE
feat: update group permissions

### PR DESCRIPTION
**Pull request by Flemmli97**
**What this PR does** 📖

- Renamed SetGroupName to EditGroupInfo
- Added RemoveParticipants permission. If someone has this they can remove a user from a group
- EditGroupInfo now allows: Renaming group name + description, changing group icon and banner
- Changed no permission error from PublicKeyInvalid to Unauthorized

- Updates the tests which only used direct conversation instead of group

### CODE REVIEW

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 1]
```diff
@@ -2340,14 +2363,20 @@ impl ConversationTask {
 
     pub async fn set_description(&mut self, desc: Option<&str>) -> Result<(), Error> {
         let conversation_id = self.conversation_id;
-        let own_did = &self.identity.did_key();
-
         if self.document.conversation_type() == ConversationType::Group {
-            let Some(creator) = self.document.creator.as_ref() else {
+            let Some(creator) = self.document.creator.clone() else {
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:2367]: Clone would be redundant here

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 2]
```diff
@@ -2157,12 +2166,19 @@ impl ConversationTask {
...
+
+        let own_did = &self.identity.did_key();
 
-        if self.document.conversation_type() == ConversationType::Group
-            && !matches!(self.document.creator.as_ref(), Some(creator) if own_did.eq(creator))
+        if !&self
+            .document
+            .permissions
+            .has_permission(own_did, GroupPermission::EditGroupInfo)
+            && creator.ne(own_did)
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:2179]: We need to have a check here to determine if the conversation is direct or group.

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 3]
```diff
@@ -2293,12 +2309,19 @@ impl ConversationTask {
...
 
-        if self.document.conversation_type() == ConversationType::Group
-            && !matches!(self.document.creator.as_ref(), Some(creator) if own_did.eq(creator))
+        let own_did = &self.identity.did_key();
+
+        if !&self
+            .document
+            .permissions
+            .has_permission(own_did, GroupPermission::EditGroupInfo)
+            && creator.ne(own_did)
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:2322]: ditto

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 4]
```diff
@@ -1931,6 +1931,15 @@ impl ConversationTask {
 
         let own_did = &self.identity.did_key();
 
+        if !self
+            .document
+            .permissions
+            .has_permission(own_did, GroupPermission::RemoveParticipants)
+            && creator.ne(own_did)
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:1938]: we should check for the creator first before checking the permission

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 5]
```diff
@@ -3482,6 +3516,14 @@ async fn message_event(
...
                 }
                 ConversationUpdateKind::AddedIcon | ConversationUpdateKind::RemovedIcon => {
+                    if !this.document.creator.as_ref().is_some_and(|c| c == sender)
+                        && !this
+                            .document
+                            .permissions
+                            .has_permission(sender, GroupPermission::EditGroupInfo)
+                    {
+                        return Err(Error::Unauthorized);
+                    }
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:3526]: A check should be added to determine if the conversation is a group before checking the permission

File: extensions/warp-ipfs/src/store/message/task.rs [hunk 6]
```diff
@@ -3493,6 +3535,14 @@ async fn message_event(
...
 
                 ConversationUpdateKind::AddedBanner | ConversationUpdateKind::RemovedBanner => {
+                    if !this.document.creator.as_ref().is_some_and(|c| c == sender)
+                        && !this
+                            .document
+                            .permissions
+                            .has_permission(sender, GroupPermission::EditGroupInfo)
+                    {
+                        return Err(Error::Unauthorized);
+                    }
```
Comment (dariusc93) [inline extensions/warp-ipfs/src/store/message/task.rs:3545]: ditto

File: warp/src/raygun/mod.rs [hunk 7]
```diff
@@ -747,7 +747,12 @@ impl Conversation {
 #[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq, Hash)]
 pub enum GroupPermission {
     AddParticipants,
-    SetGroupName,
+    RemoveParticipants,
+    EditGroupInfo,
+}
+
+impl GroupPermission {
+    pub const VALUES: [Self; 3] = [Self::AddParticipants, Self::RemoveParticipants, Self::EditGroupInfo];
```
Comment (dariusc93) [inline warp/src/raygun/mod.rs:755]: might be better to create a function called `values` that returns `Vec<GroupPermission>`

**Review by dariusc93 — COMMENTED** [review]
Left some comments. Not sure I like or completely agree with moving multiple functionalities under a single permission here (eg changing/updating name, images, etc).

**Comment by Flemmli97** [comment]
> Left some comments. Not sure I like or completely agree with moving multiple functionalities under a single permission here (eg changing/updating name, images, etc).

was talking to hugo about that. and we decided for name, description, icon and banner we would use one permission

File: extensions/warp-ipfs/src/store/message/task.rs (hunk 2 above)
Comment (Flemmli97) [reply extensions/warp-ipfs/src/store/message/task.rs:2179]: ok. wasnt sure cause previously if it were a direct one it just let it pass through

**Comment by dariusc93** [comment]
> was talking to hugo about that. and we decided for name, description, icon and banner we would use one permission

Not sure if it should be that narrow imho

**Review by ashneverdawn — CHANGES_REQUESTED** [review]

**Review by ashneverdawn — APPROVED** [review]

File: warp/src/raygun/mod.rs [hunk 8]
```diff
@@ -880,7 +880,20 @@ impl Conversation {
 #[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq, Hash)]
 pub enum GroupPermission {
     AddParticipants,
-    SetGroupName,
+    RemoveParticipants,
+    EditGroupInfo,
+    EditGroupVisual,
```
Comment (dariusc93) [inline warp/src/raygun/mod.rs:885]: I assume this is for the conversation images? If so lets name it `EditGroupImages`.

File: extensions/warp-ipfs/examples/messenger.rs [hunk 9]
```diff
@@ -528,8 +528,8 @@ async fn main() -> anyhow::Result<()> {
                                 Ok(conversation) => {
                                     let mut permissions = GroupPermissions::new();
                                     if open {
-                                        for did in conversation.recipients().to_owned() {
-                                            permissions.insert(did, vec![GroupPermission::AddParticipants, GroupPermission::SetGroupName].into_iter().collect());
+                                        for did in conversation.recipients()).to_owned() {
```
Comment (dariusc93) [inline extensions/warp-ipfs/examples/messenger.rs:531]:
```suggestion
                                        for did in conversation.recipients().to_owned() {
```

**Comment by dariusc93** [comment]
Can you correct the test?

**Pull request merged**